packages/block-editor/src/components/inserter/search-items.js

```diff
@@ -44,94 +43,98 @@ const removeMatchingTerms = ( unmatchedTerms, unprocessedTerms ) => {
  * @param {string} searchTerm Search term.
- *
- * @return {Array}             Filtered item list.
+ * @param {Object} config     Search Config.
+ * @return {Array}            Filtered item list.
  */
-export const searchItems = ( items, categories, collections, searchTerm ) => {
+export const searchItems = ( items, searchTerm, config = {} ) => {
 	const normalizedSearchTerms = normalizeSearchTerm( searchTerm );
-
 	if ( normalizedSearchTerms.length === 0 ) {
 		return items;
 	}
 
-	return items
-		.filter(
-			( { name, title, category, keywords = [], variations = [] } ) => {
-				let unmatchedTerms = removeMatchingTerms(
-					normalizedSearchTerms,
-					title
-				);
-
-				if ( unmatchedTerms.length === 0 ) {
-					return true;
-				}
-
-				unmatchedTerms = removeMatchingTerms(
-					unmatchedTerms,
-					keywords.join( ' ' )
-				);
-
-				if ( unmatchedTerms.length === 0 ) {
-					return true;
-				}
-
-				unmatchedTerms = removeMatchingTerms(
-					unmatchedTerms,
-					get( find( categories, { slug: category } ), [ 'title' ] )
-				);
-
-				const itemCollection = collections[ name.split( '/' )[ 0 ] ];
-				if ( itemCollection ) {
-					unmatchedTerms = removeMatchingTerms(
-						unmatchedTerms,
-						itemCollection.title
-					);
-				}
-
-				if ( unmatchedTerms.length === 0 ) {
-					return true;
-				}
-
-				unmatchedTerms = removeMatchingTerms(
-					unmatchedTerms,
-					variations
-						.map( ( variation ) => variation.title )
-						.join( ' ' )
-				);
-
-				return unmatchedTerms.length === 0;
-			}
-		)
-		.map( ( item ) => {
-			if ( isEmpty( item.variations ) ) {
-				return item;
-			}
-
-			const matchedVariations = item.variations.filter( ( variation ) => {
-				return (
-					intersectionWith(
-						normalizedSearchTerms,
-						normalizeSearchTerm( variation.title ),
-						( termToMatch, labelTerm ) =>
-							labelTerm.includes( termToMatch )
-					).length > 0
-				);
-			} );
-			// When no partterns matched, fallback to all variations.
-			if ( isEmpty( matchedVariations ) ) {
-				return item;
-			}
-
-			return {
-				...item,
-				variations: matchedVariations,
-			};
-		} );
+	const defaultGetTitle = ( item ) => item.title;
+	const defaultGetKeywords = ( item ) => item.keywords || [];
+	const defaultGetCategory = ( item ) => item.category;
+	const defaultGetCollection = () => null;
+	const defaultGetVariations = () => [];
+	const {
+		getTitle = defaultGetTitle,
+		getKeywords = defaultGetKeywords,
+		getCategory = defaultGetCategory,
+		getCollection = defaultGetCollection,
+		getVariations = defaultGetVariations,
+	} = config;
+
+	return items.filter( ( item ) => {
+		const title = getTitle( item );
+		const keywords = getKeywords( item ).join( ' ' );
+		const category = getCategory( item );
+		const collection = getCollection( item );
+		const variations = getVariations( item ).join( ' ' );
+
+		const terms = [
+			title,
+			keywords,
+			category,
+			collection,
+			variations,
+		].join( ' ' );
```

Comment:
Rather than calling `join` on individual items, you can run `flat` here:

```suggestion
		].flat().join( ' ' );
```